### DISCUSSION TITLE
Is it possible to use equivalent of magick compare with this wasm?

### DISCUSSION THREAD

**Discussion by rgembalik** [dlemstra/magick-wasm#169] [Q&A] [answered]
I saw that the `IMagickImage.compare` is implemented, but it returns a numeric value. Is it possible to get a diff image as it is returned with https://imagemagick.org/script/compare.php?

**Comment by dlemstra** [comment]
This is possible with the underlying API but I have not exposed this yet in the typescript api. I can add this for you but I could use your help with how we should name this new method and it's signature.

```typescript
// This is the current method and we should keep this.
compare(image: IMagickImage, metric: ErrorMetric): number;

// And add an extra overload
compareSomething(image: IMagickImage, metric: ErrorMetric, func: (something: ISomething) => TReturnType): TReturnType;

interface ISomething
{
   distortion: number;
   something: IMagickImage;
}
```

Can you help me with names for the somethings?

**Reply by rgembalik** [reply]
Sure. Based on how I was searching for it in the sources, I would use it like this:

```typescript
compareVisual(image: IMagickImage, metric: ErrorMetric): ICompareResult;

interface ICompareResult
{
   distortion: number;
   diffImage: IMagickImage;
}
```

And I totally agree on the separate method - there is no point in confusing the output of the existing method with multiple return types.

If you decide to implement it, give me a shout once this is done - I'd be happy to see how it was written and based on that, in the future I could try to create PR on my own.

**Reply by dlemstra** [reply]
I have updated my comment and changed the signature a bit because the `diffImage` should only be available inside a function because it is using native memory and that needs to be cleaned up when you are done with the image. I do like the `ICompareResult` name so I will be going to use that. And I will let you know how this change was made.

**Reply by dlemstra** [reply]
I just pushed this patch that added support for this: https://github.com/dlemstra/magick-wasm/commit/1b68ec5d68d37bf769142317dcdcdb5624c339c9. I will be going to add an extra overload with some extra settings tomorrow or this weekend.

This patch also included some extra work because this was the first time that I needed a pointer for a double.

**Reply by rgembalik** [reply]
Awesome! Thank you for the update and generally for such a quick response. 

This looks great. I'd be happy to test it once it's released, even as beta (I tried installation from GH, but since I cannot build it w/o `magick-native` I figured it's best to wait).

**Reply by dlemstra** [reply, marked as answer]
I just published a new release that includes the changes that I made. Thanks for helping out.
